pkg/operation/botanist/containerruntime.go

```diff
@@ -158,9 +125,51 @@ func (b *Botanist) WaitUntilContainerRuntimeResourcesReady(ctx context.Context)
 	return flow.ParallelExitOnError(fns...)(ctx)
 }
 
-// DeleteContainerRuntimeResources deletes all container runtime resources from the Shoot namespace in the Seed.
-func (b *Botanist) DeleteContainerRuntimeResources(ctx context.Context) error {
-	return b.K8sSeedClient.Client().DeleteAllOf(ctx, &extensionsv1alpha1.ContainerRuntime{}, client.InNamespace(b.Shoot.SeedNamespace))
+// DeleteStaleContainerRuntimeResources deletes unused container runtime resources from the shoot namespace in the seed.
+func (b *Botanist) DeleteStaleContainerRuntimeResources(ctx context.Context) error {
+	wantedContainerRuntimeTypes := sets.NewString()
+	for _, worker := range b.Shoot.Info.Spec.Provider.Workers {
+		if worker.CRI != nil {
+			for _, containerRuntime := range worker.CRI.ContainerRuntimes {
+				wantedContainerRuntimeTypes.Insert(containerRuntime.Type)
+			}
+		}
+	}
+	return b.deleteContainerRuntimeResources(ctx, wantedContainerRuntimeTypes)
+}
+
+// DeleteAllContainerRuntimeResources deletes all container runtime resources from the Shoot namespace in the Seed.
+func (b *Botanist) DeleteAllContainerRuntimeResources(ctx context.Context) error {
+	return b.deleteContainerRuntimeResources(ctx, sets.NewString())
+}
+
+func (b *Botanist) deleteContainerRuntimeResources(ctx context.Context, wantedContainerRuntimeTypes sets.String) error {
+	deployedContainerRuntimes := &extensionsv1alpha1.ContainerRuntimeList{}
+	if err := b.K8sSeedClient.Client().List(ctx, deployedContainerRuntimes, client.InNamespace(b.Shoot.SeedNamespace)); err != nil {
+		return err
+	}
+
+	fns := make([]flow.TaskFn, 0, meta.LenList(deployedContainerRuntimes))
+	for _, deployedContainerRuntime := range deployedContainerRuntimes.Items {
+		if !wantedContainerRuntimeTypes.Has(deployedContainerRuntime.Spec.Type) {
+			toDelete := &extensionsv1alpha1.ContainerRuntime{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      deployedContainerRuntime.Name,
+					Namespace: deployedContainerRuntime.Namespace,
+				},
+			}
+
+			fns = append(fns, func(ctx context.Context) error {
+				if err := common.ConfirmDeletion(ctx, b.K8sSeedClient.Client(), toDelete); err != nil {
```

Comment:
Is the ContainerRuntime resource being checked by the SeedAdmission plugin to contain that deletion annotation?

Reply:
Yes, the seed admission controller checks all CRDs having a special annotation, and the annotation was already added by @nimrodoron's PR. https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/templates/extensions/crd-containerruntime.yaml#L6